core/src/main/java/io/grpc/transport/WritableBufferAllocator.java

```diff
@@ -32,13 +32,14 @@
 package io.grpc.transport;
 
 /**
- * The preferred way of creating a {@link WritableBuffer}.
+ * An allocator of buffers provided by the transport implementation to {@link MessageFramer} so
+ * it can send chunks of data to the transport in a form that the transport can directly serialize.
  */
 public interface WritableBufferAllocator {
 
   /**
-   * Allocate a new {@link WritableBuffer} that can hold
-   * {@code capacity} bytes.
+   * Request a new {@link WritableBuffer} with the given {@code capacity}. The allocator is
+   * free to return a buffer with a greater or lesser capacity.
```

Comment:
So there are a number of step-functions in the pooled buffer allocator that make this choice interesting and I suspect we'll do a couple of rounds on this and we probably want benchmarks to dictate whether it's better to to send larger buffers into the flow-controller or smaller.

I'm running Netty's benchmarks to inform the decision for the lower-bound but the choice for the upper-bound is more interesting...

- Picking an upper-bound that is guaranteed to fit within the arena does not guarantee that we get a pooled buffer (the arena might be exhausted)
- Taking buffers out of the arena as opposed to the buffer pool (> 32k) means larger locks so there is a tradeoff between the cost of those locks and the cost of main-heap allocation.
- If we do enforce an upper bound we will consume aggressively from the pool. A single very large object could exhaust the pool and starve smaller allocations. This is the opposite of what we want as the pool is designed to optimize smaller allocs.

I'm inclined for the moment to set an upper-bound but in truth we need benchmarking to determine (a) what it should be and (b) if we really need it at all.